File: java/src/main/java/br/com/emendes/problems/ReverseLinkedListII.java
package br.com.emendes.problems;

import br.com.emendes.problems.util.ListNode;

/**
 * Dado a head de uma linked list e dois inteiros left e right onde left <= right,
 * inverta os nodes da lista da posição esquerda para a posição direita.
 * E retorne a linked list.<br>
 * <br>
 * Restrições:<br>
 * O número de nodes na lista é n.<br>
 * 1 <= n <= 500<br>
 * -500 <= Node.val <= 500<br>
 * 1 <= left <= right <= n
 */
public class ReverseLinkedListII {

  public ListNode reverseBetween(ListNode head, int left, int right) {
    if (left == right) return head;
    ListNode dummy = new ListNode(0, head);
    ListNode leftLeft = dummy;
    ListNode nodeLeft = head;
    for (int i = 1; i < left; i++) {
      leftLeft = nodeLeft;
      nodeLeft = nodeLeft.next;
    }
    ListNode current = nodeLeft.next;
    ListNode previous = nodeLeft;
    ListNode next;
    left++;
    while (left <= right) {
      next = current.next;
      current.next = previous;
      previous = current;
      current = next;
      left++;
    }
    nodeLeft.next = current;
    leftLeft.next = previous;

    return dummy.next;
  }

}

File: javascript/problems/ReverseLinkedListII.js
/**
 * Dado o head de uma linked list e dois inteiro left e right onde left <= right,
 * reverta os nodes da lista da posição left até a posição right, e retorne a lista revertida.
 *
 * Restrições:
 *
 * - o número de nodes da lista é n.
 *
 * - 1 <= n <= 500
 *
 * - -500 <= Node.val <= 500
 *
 * - 1 <= left <= right <= n
 */
const reverseBetween = function (head, left, right) {
  if (left === right) return head;
  const dummy = new ListNode(-1, head);
  let prev = dummy;
  for (let i = 0; i < left - 1; i++) prev = prev.next;

  const tail = prev.next;
  for (let i = 0; i < right - left; i++) {
    const cache = tail.next;
    tail.next = cache.next;
    cache.next = prev.next;
    prev.next = cache;
  }
  return dummy.next;
};

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

function generateListNode(values) {
  const dummy = new ListNode(-1);
  let node = dummy;

  for (const val of values) {
    node.next = new ListNode(val);
    node = node.next;
  }
  return dummy.next;
}
